pkg/virt-api/webhooks/validating-webhook/admitters/pod-eviction-admitter.go

```diff
@@ -24,15 +25,17 @@ type PodEvictionAdmitter struct {
 }
 
 func (admitter *PodEvictionAdmitter) Admit(ar *admissionv1.AdmissionReview) *admissionv1.AdmissionResponse {
-	launcher, err := admitter.VirtClient.CoreV1().Pods(ar.Request.Namespace).Get(context.Background(), ar.Request.Name, metav1.GetOptions{})
+	pod, err := admitter.VirtClient.CoreV1().Pods(ar.Request.Namespace).Get(context.Background(), ar.Request.Name, metav1.GetOptions{})
 	if err != nil {
 		return validating_webhooks.NewPassingAdmissionResponse()
 	}
 
-	if value, exists := launcher.GetLabels()[virtv1.AppLabel]; !exists || value != "virt-launcher" {
+	if !isVirtLauncher(pod) {
 		return validating_webhooks.NewPassingAdmissionResponse()
 	}
 
+	launcher := pod
```

Comment:
> I don't get the real added value here honestly and the launcher := pod is something that I will not encourage.

Currently, it is not possible to filter eviction requests by pod label [1][2], so unfortunately this webhook intercepts all eviction requests in the cluster - including for pods that are not virt-launchers.

Since the eviction request does not contain any data on the evicted pod other than its namespace/name, the webhook has to fetch the pod in order to understand whether its a virt-launcher or not (by reading its `kubevirt.io` label and expecting the value to be `virt-launcher`).

Only after this test, the pod could be considered a virt-launcher.

The current code called the fetched pod `launcher`, even if it is not a virt-launcher pod.
I found it to be confusing and misleading.

> The good refactoring that I'd expect here is to use the labelselector while retrieving the virt-launcher pod

AFAIK, using labelSelectors is useful when performing List operations.
In this case we perform a Get operation to get a single pod.
Could you please elaborate what do you mean in this context?

[1] https://github.com/kubernetes/kubernetes/issues/110169#issuecomment-1140512056
[2] https://kubernetes.slack.com/archives/C0EG7JC6T/p1707054818877809

Reply:
For completeness, the AdmissionReview object sent from kube-api to the webhook looks like this:
```json
{
  "kind": "AdmissionReview",
  "apiVersion": "admission.k8s.io/v1",
  "request": {
    "uid": "f80b8729-034e-4198-bcc9-1e7aca659ca1",
    "kind": {
      "group": "policy",
      "version": "v1",
      "kind": "Eviction"
    },
    "resource": {
      "group": "",
      "version": "v1",
      "resource": "pods"
    },
    "subResource": "eviction",
    "requestKind": {
      "group": "policy",
      "version": "v1",
      "kind": "Eviction"
    },
    "requestResource": {
      "group": "",
      "version": "v1",
      "resource": "pods"
    },
    "requestSubResource": "eviction",
    "name": "mypod",
    "namespace": "default",
    "operation": "CREATE",
    "userInfo": {
      "username": "kubernetes-admin",
      "groups": [
        "kubeadm:cluster-admins",
        "system:authenticated"
      ]
    },
    "object": {
      "kind": "Eviction",
      "apiVersion": "policy/v1",
      "metadata": {
        "name": "mypod",
        "namespace": "default",
        "creationTimestamp": null,
        "managedFields": [
          {
            "manager": "kubectl",
            "operation": "Update",
            "apiVersion": "policy/v1",
            "time": "2024-02-28T09:32:08Z",
            "fieldsType": "FieldsV1",
            "fieldsV1": {
              "f:deleteOptions": {}
            },
            "subresource": "eviction"
          }
        ]
      },
      "deleteOptions": {}
    },
    "oldObject": null,
    "dryRun": false,
    "options": {
      "kind": "CreateOptions",
      "apiVersion": "meta.k8s.io/v1"
    }
  }
}
```

Reply:
Sorry, maybe I misexplained what was not clear to me. The fact that we were excluding non-virtlauncher pods was clear.
I didn't understand what was the reasoning behind this commit.
I think that if a refactoring can be done here is to filter the non-virtlauncher pods serverside, with labelselector.
So something like:
```
	pods, err := admitter.VirtClient.CoreV1().Pods(ar.Request.Namespace).List(context.Background(), metav1.ListOptions{LabelSelector: fmt.Sprintf(virtv1.AppLabel + "=virt-launcher"), FieldSelector: "metadata.name=" + ar.Request.Name})
	if err != nil || pods.Size() == 0 {
		return validating_webhooks.NewPassingAdmissionResponse()
	}

	launcher := pods.Items[0]
```

WDYT?

Reply:
An eviction request is sent from its initiator, lets say kubectl, per pod [1].
Hence in this webhook, we get the AdmissionReview object derived from a single eviction request, containing only the evicted pod's namespace/name.

This is the reason a Get operation is performed - it gets a single pod by namespace/name.

IMO the List operation is:
1. More costly on kube-api's side.
2. A bit more costly on the network since more JSON is sent.
3. Is less readable.

I don't a clear advantage over the existing approach (which was unchanged by this PR).

[1] https://kubernetes.io/docs/concepts/scheduling-eviction/api-eviction/#calling-the-eviction-api

Reply:
After an offline conversation maybe here is better to continue using `Get` (thanks for the discussion).
But please, remove `launcher := pod` :)

Reply:
> After an offline conversation maybe here is better to continue using `Get` (thanks for the discussion). But please, remove `launcher := pod` :)

Could you please explain the reason behind your objection? I haven't found it in this thread.
Please note that this copy has very low cost - only a pointer is copied.

Reply:
```go
if !isVirtLauncher(launcher) {
}
```

Reply:
You can always rename the returned value from  the `Get` operation, so that it will be:
```
if !isVirtLauncher(pod) {
}
```

Reply:
Done.